src/main/scala/com/fulcrumgenomics/commons/async/Async.scala

```diff
@@ -0,0 +1,397 @@
+      }
+    }
+  }
+
+  object AsyncSink {
+    /** A counter used for thread naming in [[AsyncSink]] */
+    private val threadsCreated = new AtomicInteger(0)
+  }
+
+  /** An asynchronous wrapper for processing items of type [[T]] and process them in a separate thread using
+    * the provided `sink` method.
+    *
+    * NOTE: Any exception thrown by the [[sink]] method will be propagated back to the caller
+    * during the next available call to [[add()]] or [[close()]]. After the exception
+    * has been thrown to the caller, it is not safe to attempt further operations on the instance.
+    *
+    * NOTE: [[add()]] or [[close()]] are not thread-safe, so only there must be only one thread that calls them.
+    *
+    * @param sink the method to invoke to process an object of type [[T]].
+    * @param bufferSize the number of elements to buffer before blocking when processing the elements, or None if
+    *                   unbounded.
+    * @param source the optional source to close when this sink is closed.
+    * @tparam T the type of object to process
+    */
+  class AsyncSink[T](val sink: T => Unit, bufferSize: Option[Int] = None, val source: Option[{ def close(): Unit }] = None)
+    extends Closeable with AsyncThread {
+    bufferSize.foreach(size => require(size > 0, s"bufferSize must be greater than zero when given, found $size"))
+    private val queue: BlockingQueue[T] = bufferSize match {
+      case Some(size) => new ArrayBlockingQueue[T](size)
+      case None       => new LinkedBlockingDeque[T]()
+    }
+    private val thread = buildThread(this.getClass.getSimpleName + AsyncSink.threadsCreated.incrementAndGet)
+    private val outer = this
+
+    // Start the thread after creating the queue
+    this.thread.start()
+
+    /** Adds the item to the queue to be processed by [[sink]].
+      *
+      * @param item the item to queue
+      */
+    def add(item: T): Unit = {
+      if (this.isDone.get) throw new RuntimeException("Attempt to add record to closed sink.")
+      checkAndRethrow()
+      tryAndModifyInterruptedException("Interrupted queueing item.") { this.queue.put(item) }
+      checkAndRethrow()
+    }
+
+    /** Attempts to finish draining the queue and then calls `close()` on the source to allow implementation
+      * to do any one time clean up.
+      */
+    def close(): Unit = {
+      checkAndRethrow()
+      if (!this.isDone.getAndSet(true)) {
+        tryAndModifyInterruptedException("Interrupted waiting on sink thread.") { this.thread.join() }
+        // The queue should be empty but if it's not, we'll drain it here to protect against any lost data.
+        // There's no need to timeout on poll because take is called only when queue is not empty and
+        // at this point the thread is definitely dead and no-one is removing items from the queue.
+        // The item pulled will never be null (same reasoning).
+        while (!this.queue.isEmpty) {
```

Comment:
These suspenders don't go with this belt!  Why not just `require(this.queue.isEmpty)` here?